### PR TITLE
feat: add sign method to transaction model

### PR DESCRIPTION
**Pull request by r4mmer**
### Acceptance Criteria
- Create a sign method on the transaction model

### Security Checklist
- [ ] Make sure you do not include new dependencies in the project unless strictly necessary and do not include dev-dependencies as production ones. More dependencies increase the possibility of one of them being hijacked and affecting us.

### CODE REVIEW

**Review by tuliomir — COMMENTED** [review]
Since you're refactoring the signature flow, maybe it would be a good idea to centralize the signing code too.

Currently we use the `crypto.ECDSA.sign` in four places:
- models/transaction.ts - `sign`
- new/wallet.js - `getAllSignatures` _( uses `.toString()` instead of `.toDER()` )_
- transaction.js - `getSignature`
- wallet/wallet.ts - `getInputData`

It seems to me that we can be confused later with all these different places to sign transaction parts. In this case, the static `Transaction.sign` method could be moved to a _util_ file and then consumed from all these four places.

Do you think this would be a good idea, and a relevant addition to the scope of this PR?

**Review by tuliomir — APPROVED** [review]

File: src/utils/transaction.ts [hunk 1]
```diff
@@ -5,11 +5,29 @@
...
-
 import { Utxo } from '../wallet/types';
 import { UtxoError } from '../errors';
+import {crypto as cryptoBL, PrivateKey} from 'bitcore-lib'
 
 const transaction = {
+  /**
+   * Get the signature from the dataToSignHash for a private key
+   *
+   * @param {Buffer} dataToSignHash sha256 of a transaction's dataToSign.
```
Comment (pedroferreira1) [inline src/utils/transaction.ts:16]: This is not only the sha256 of data to sign, we need to do a little bit more. You can say that it's the hash of `dataToSign`

File: src/models/transaction.ts [hunk 2]
```diff
@@ -92,6 +93,32 @@ class Transaction {
...
 
+  /**
+   * Get the signature from the dataToSignHash for a private key
+   *
+   * @return {Buffer}
+   * @memberof Transaction
+   * @static
+   *
+   */
+  static sign(dataToSignHash: Buffer, privateKey: PrivateKey): Buffer {
```
Comment (pedroferreira1) [inline src/models/transaction.ts:104]: Why do you have this method as well? You could use only the other one, the static method is the same as the utils method, right?

File: src/models/transaction.ts [hunk 3]
```diff
@@ -92,6 +93,32 @@ class Transaction {
...
+   * Get the signature of this transaction for the given private key
+   *
+   * @param {PrivateKey}
+   *
+   * @return {Buffer}
+   * @memberof Transaction
+   * @inner
+   *
+   */
+  sign(privateKey: PrivateKey): Buffer {
```
Comment (pedroferreira1) [inline src/models/transaction.ts:118]: I would change the method name, this is `getSignature`, you are not signing the transaction.

File: src/new/wallet.js [hunk 4]
```diff
@@ -359,15 +358,11 @@ class HathorWallet extends EventEmitter {
...
       const derivedKey = key.deriveNonCompliantChild(addressIndex);
       const privateKey = derivedKey.privateKey;
 
-      // derive key to address index
-      const sig = crypto.ECDSA.sign(hash, privateKey, 'little').set({
-        nhashtype: crypto.Signature.SIGHASH_ALL
-      });
-
-      signatures[index] = sig.toString();
+      signatures[index] = tx.sign(privateKey).toString('hex');
```
Comment (pedroferreira1) [inline src/new/wallet.js:365]: The old code here was doing `sig.toString()` before doing `sig.toDER()`. The new code is doing `sig.toDER().toString('hex')`, is it the same?

File: src/utils/transaction.ts [hunk 5]
```diff
@@ -5,11 +5,29 @@
...
+   * @param {PrivateKey} privateKey Signing key.
+   *
+   * @returns {Buffer}
+   *
+   * @memberof transaction
+   * @inner
+   */
+  sign(dataToSignHash: Buffer, privateKey: PrivateKey): Buffer {
```
Comment (pedroferreira1) [inline src/utils/transaction.ts:24]: I think the same `getSignature` is better for this method, you are not signing anything. What do you think?

File: src/new/wallet.js (hunk 4 above)
Comment (r4mmer) [reply src/new/wallet.js:365]: bitcore-lib's [`Signature.toString()`](https://github.com/bitpay/bitcore/blob/master/packages/bitcore-lib/lib/crypto/signature.js#L189-L192) uses `toDER()` then `toString('hex')` internally.
This means we can just use `toString('hex')` here.

File: src/models/transaction.ts (hunk 2 above)
Comment (r4mmer) [reply src/models/transaction.ts:104]: Yes, it´s exactly the same method. Just kept it here for convenience, do you think we should remove to keep this cleaner?

File: src/models/transaction.ts (hunk 3 above)
Comment (r4mmer) [reply src/models/transaction.ts:118]: Done

File: src/utils/transaction.ts (hunk 5 above)
Comment (r4mmer) [reply src/utils/transaction.ts:24]: Done.

File: src/models/transaction.ts (hunk 2 above)
Comment (pedroferreira1) [reply src/models/transaction.ts]: I feel we could remove it and keep only the other one.

**Review by pedroferreira1 — APPROVED** [review]

**Pull request merged**